cli/main.c

```diff
@@ -1347,6 +1348,19 @@ static int fw_update(int argc, char **argv)
 
 	argconfig_parse(argc, argv, desc, opts, &cfg, sizeof(cfg));
 
+	ret = switchtec_get_device_info(cfg.dev, &phase_id, NULL, NULL);
+	if (ret < 0) {
+		switchtec_fw_perror("firmware update", ret);
+		return ret;
+	}
+	if (phase_id == SWITCHTEC_BOOT_PHASE_BL1) {
+		fprintf(stderr,
+			"This command is only available in BL2 or Main Firmware!\n");
+		fprintf(stderr,
+			"Use 'fw-transfer' instead to download a BL2 image.\n");
```

Comment:
'mfg fw-transfer', to 'transfer' instead of download?

Reply:
Wait why can't fw_update just do the right thing for a BL2 image? Why do we need to create a new command for it?

Reply:
Oh this message is printed in BL1 mode. 

BL1 (Bootloader 1) is implemented in an on-chip boot ROM and has limited capabilities, hence a different command for transferring image in BL1.

Reply:
Ok, so you require a different MRPC command/procedure... this does not at all justify a different user API and command line interface. This just means everyone using it, be it with a shell script or via the library have to have an if (BL1) case for no good reason.

Reply:
The BL1 'fw-transfer' command has a quite different behavior from 'fw-update' in BL2 or MAIN. 

With 'fw-transfer', you can only send BL2 image to device RAM, no writing to flash or anything. 
'fw-update' supports downloading any (BL2, MAIN, CFG) image to flash.

This line was previously added at request from our APPS team, so users trying to update FW image in BL1 will get a bit more information about what else they can try, instead of just a cold error "command not available."

The 'fw-transfer' command in 'mfg' submenu will be checked in later.

If this line creates more confusion, then maybe we can remove it.

Reply:
I still don't see why the existing fw update API can't just see that we're in BL1 and do the right thing.... If the CLI prints some warnings or something saying we're in BL1 and therefore the new image is only stored in ram, then that seems acceptable. But the API should still be simple and take care of these details for the user.

Reply:
A little background info:
_One use case for mfg commands is recovering from boot failure due to corrupt partitions. In this case, user can set a jumper pin on the board, and the device will boot into BL1 (always available since it's in on-chip ROM). 

From BL1, user can do 'fw-transfer' to send a BL2 image to RAM, then do 'fw-execute' to run the image ('fw-transfer' does not automatically execute the image). The device will boot into BL2 phase, and then user can do 'fw-update' to update corrupted partitions._

Previously we have 'mfg fw-transfer' for BL1, and 'fw-update' for BL2/MAIN. From the above discussion, it seems we can move 'fw-transfer' and 'fw-execute' command from **mfg** submenu to main menu, and merge 'fw-transfer' into 'fw-update': print a message that this command is running in BL1 so the image is sent to RAM instead of flash. 

Note that in BL1, user still need to send 'fw-execute' command to execute the image.

@lsgunth let me know what you think.

Reply:
Yes, merge fw-transfer and fw-update...

I think fw-execute could still be in the mfg menu... Though I don't feel that strongly about it. You must be very clear that fw-execute only applies to uploads during BL1, unless you have a way to support having it work after a regular firmware update.

Also, fw-execute sounds super dangerous... How does this interact with the host OS? Seems like it would need to reinitialize everything...

Reply:
The fw-transfer transfers a BL2 image into the chip ram, and the following fw-execute jumps to the tranfered BL2 image and executes. Thereafter, the BL2 fw starts to run in a while(1) loop fashion with limited commands supported aiming to do some recovery related stuff that user may ask it to do by MRPC commands. 

I think the fw-transfer and fw-execute are two steps of one operation (run a BL2 fw), therefore it makes more sense to have them grouped in the same mfg sub-menu. For the API, I'm good if we have them implemented in the current fw update APIs.

So I'd prefer we have the fw-transfer as a seperate mfg command with only needed option (none?), without being mixed with the fw-update which have many other options which are meaningless to the transfer operation. Meanwhile we have the same API supporting both download and transfer.

Reply:
I disagree. fw-transfer and fw-update do exactly the same thing (push firmware to the device). Having two different commands is confusing to the user. It's fine to ignore or warn if they specify arguments that don't apply in the situation. The API will be almost exactly the same and would also be confusing on which is used when. Better to avoid the extra code and documentation burden and just adapt the existing command.

Reply:
Here is a brief summary, based on the current implementation:
- at API level, there is only one function _switchtec_fw_write_file()_, both _fw-transfer_ and _fw-update_ use this function to send FW image to device
- at CLI command level, @lsgunth prefer combining _fw-transfer_ into _fw-update_, while @kelvin-cao wants to keep them separate.

Based on what I see so far we all like the idea of keeping the same API function _switchtec_fw_write_file()_ for BL1/2 & MAIN. So at this point we can say this API definition is decided. 

At CLI command level, I agree with Kelvin that the _fw-transfer_ command should be grouped with _fw-execute_, in mfg submenu. From a user perspective, it would be weird to do FW transfer using command from Main Menu, then do _fw-execute_ from mfg submenu. 

Keeping these 2 commands separate would mean some duplicate code at CLI level, but the _fw-transfer_ command is much simpler so there should not be that much of duplicate code.

@lsgunth what do you think?

Reply:
If you absolutely must have a separate CLI command I would expect *very* good documentation in the 'switchtec help' for both commands. It must explain when each command is the correct one to use and the difference between them. You must also define your terms. Saying 'fw-transfer' can only be used in "BL1" will not be sufficient because most people will not know what BL1 is or when it applies.  

This would be a case where the description in the command should not match the description in the list because the former will need to be quite long and detailed.

Reply:
@lsgunth code updated with more information in 'fw-update' descriptions. Please take a look.